compiler/src/dotty/tools/dotc/interactive/Completion.scala

```diff
@@ -135,161 +135,166 @@ object Completion {
+  /**
+   * A description for completion result that represents `symbols`.
+   *
+   * If `denots` contains a single denotation, show its full name in case it's a type, or its type if
+   * it's a term.
+   *
+   * When there are multiple denotations, show their kinds.
+   */
+  private def description(denots: List[SingleDenotation])(using Context): String =
+    denots match {
+      case denot :: Nil =>
+        if (denot.isType) denot.symbol.showFullName
+        else denot.info.widenTermRefExpr.show
 
-    private val completions = new RenameAwareScope
+      case denot :: _ =>
+        denots.map(den => ctx.printer.kindString(den.symbol)).distinct.mkString("", " and ", s" ${denot.symbol.name.stripModuleClassSuffix.show}")
 
-    /**
-     * Return the list of symbols that should be included in completion results.
-     *
-     * If several symbols share the same name, the type symbols appear before term symbols inside
-     * the same `Completion`.
-     */
-    def getCompletions(using Context): List[Completion] = {
-      val nameToSymbols = completions.mappings.toList
-      nameToSymbols.map { case (name, symbols) =>
-        val typesFirst = symbols.sortWith((s1, s2) => s1.isType && !s2.isType)
-        val desc = description(typesFirst)
-        Completion(name.show, desc, typesFirst)
-      }
+      case Nil =>
+        ""
     }
 
-    /**
-     * A description for completion result that represents `symbols`.
-     *
-     * If `symbols` contains a single symbol, show its full name in case it's a type, or its type if
-     * it's a term.
-     *
-     * When there are multiple symbols, show their kinds.
+  private class CompletionBuffer(val mode: Mode, val prefix: String, pos: SourcePosition) {
+    /** Completions for terms and types that are currently in scope:
+     *  the members of the current class and the symbols that have been imported, recursively adding completions from outer scopes
      */
-    private def description(symbols: List[Symbol])(using Context): String =
-      symbols match {
-        case sym :: Nil =>
-          if (sym.isType) sym.showFullName
-          else sym.info.widenTermRefExpr.show
-
-        case sym :: _ =>
-          symbols.map(ctx.printer.kindString).mkString("", " and ", s" ${sym.name.show}")
-
-        case Nil =>
-          ""
+    def scopeCompletions(using Context): CompletionScope = {
+      extension[A, K](elems: Seq[A])
+        def orderedGroupBy(f: A => K): Seq[(K, Seq[A])] =
+          val keys = elems.map(f).distinct
+          val grouped = elems.groupBy(f)
+          keys.map(key => key -> grouped(key))
+
+      val grouped = ctx.outersIterator.toList.reverse.orderedGroupBy(_.owner).filter(_._1.exists)
+      val imported = grouped.map { (owner, contexts) =>
+        contexts.collect { case context if context.isImportContext =>
+          importedCompletions(using context)
+        }.foldLeft(CompletionScope.empty)(_.mergeDiscardingAmbiguities(_))
```

Comment:
Actually the previous implementation, which used recursion, couldn't handle some of the use cases well. It's important that imported completions get shadowed by members completions to match how the compiler sees symbols in such situations. I tried to be quite explicit when ambiguities are resolved by shadowing and when using ambiguous symbols would break compilation

Reply:
> Actually the previous implementation, which used recursion, couldn't handle some of the use cases well.

That is something interesting to know -- conceptually calling the higher-order functions can be rewritten without fusing and inlining. When we have complex chaining with group/ordering, I think writing the code in a plain style will be more maintainable.

Reply:
Not really sure what you mean by "plain style" here but using only one `grouped.map {` block would probably require returning some tricky tuples inside and then unzipping the result to achieve the same semantics

Reply:
By plain style, I mean try to avoid using high-order functions like `map/fold/filter/group/etc`, and use `while` or recurse methods instead.

Reply:
I still feel uneasy with the code inside `scopeCompletions`. There are too many chaining of high-order functions. A recursive method or while-loop will be much better in readability and performance.

Could you try to rewrite the code in a plain style to encode the algorithm here? For reference, in the compiler there are many places we use recursion and while-loops to deal with the contex:

https://github.com/lampepfl/dotty/blob/14927459330da0e26f591bd73d330fa6bb198674/compiler/src/dotty/tools/dotc/typer/ImportInfo.scala#L224-L225